### PR TITLE
And version info to modules

### PR DESCRIPTION
**Pull request by awortman-fastly**
also add a check in lucet-runtime that modules loaded have version
version present, and that the module's version matches the runtime,
rejecting modules that do not

edit: the PR does fewer things, so I fixed the title and message to match

### CODE REVIEW

**Comment by awortman-fastly** [comment]
this is a sketch of what's been kicking around in my head, wrote it out partially motivated by #291 today - this also adds a good place to include further codegen quirks like "this module uses pinned heaps". The lack of a good place for that kind of setting is why #273 remains a draft, so I figure this is a good time.

File: lucet-module/src/version_info.rs [hunk 1]
```diff
@@ -0,0 +1,112 @@
...
+pub struct VersionInfo {
+    major: u8,
+    minor: u8,
+    patch: u8,
+    reserved: u8,
+    pub capabilities: Capabilities,
+}
+
+impl VersionInfo {
+    pub fn current() -> Self {
```
Comment (awortman-fastly) [inline lucet-module/src/version_info.rs]: My expectation is that `lucet-runtime`'s version check would always use `VersionInfo::current`, sets supported capabilities, and then does a comparison with whatever we loaded. That comparison function is conspicuously absent here; I think we tend to know what will be breaking change to a module's ABI, and can just bump these fields accordingly, but that's where I'd really like thoughts

File: lucet-module/src/version_info.rs [hunk 2]
```diff
@@ -0,0 +1,92 @@
...
+/// how the module can be loaded, if so requested. The information here describes implementation
+/// details in runtime support for `lucetc`-produced modules, and nothing higher level.
+#[repr(C)]
+#[derive(Debug, Clone, PartialEq, Eq)]
+pub struct VersionInfo {
+    major: u16,
+    minor: u16,
+    patch: u16,
+    reserved: u16,
+    version_hash: [u8; 8],
```
Comment (acfoltzer) [inline lucet-module/src/version_info.rs]: Should note which portion of the hash this is

File: lucetc/src/output.rs [hunk 3]
```diff
@@ -136,10 +136,15 @@ fn write_module(
     function_manifest_len: usize,
     obj: &mut Artifact,
 ) -> Result<(), Error> {
-    let mut native_data = Cursor::new(Vec::with_capacity(NATIVE_POINTER_SIZE * 4));
+    let mut native_data = Cursor::new(Vec::with_capacity(16 + NATIVE_POINTER_SIZE * 4));
```
Comment (acfoltzer) [inline lucetc/src/output.rs]: Does this have to be a magic number or can we use `size_of`?

File: lucet-module/src/version_info.rs (hunk 2 above)
Comment (awortman-fastly) [reply lucet-module/src/version_info.rs]: you mean that it's the first eight ascii characters of the git hash?

File: lucetc/src/output.rs (hunk 3 above)
Comment (awortman-fastly) [reply lucetc/src/output.rs]: good point! the entire `with_capacity` argument can and should be `mem::size_of::<SerializedModule>` now that such a type exists.

File: lucet-module/src/version_info.rs (hunk 2 above)
Comment (acfoltzer) [reply lucet-module/src/version_info.rs]: Yep

File: lucet-runtime/tests/version_checks.rs [hunk 4]
```diff
@@ -0,0 +1,16 @@
+use lucet_runtime::{DlModule, Error};
+
+#[test]
+pub fn reject_old_modules() {
+    let err = DlModule::load("./tests/version_checks/old_module.so")
```
[the remark this answers is not in the document]
Comment (iximeow) [reply lucet-runtime/tests/version_checks.rs]: Note from future-me to past-me, _what source was this module built from?_

**Pull request merged**